hack/feature-delete.sh

```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+set -e
+
+if [ "$FEATURES_ENVIRONMENT" == "" ]; then
+	echo "[ERROR]: No FEATURES_ENVIRONMENT provided"
+	exit 1
+fi
+
+if [ "$FEATURES" == "" ]; then
+	echo "[ERROR]: No FEATURES provided"
+	exit 1
+fi
+
+# expect oc to be in PATH by default
+export OC_TOOL="${OC_TOOL:-oc}"
+
+# Destroy features
+
+for feature in $FEATURES; do
+
+  feature_dir=feature-configs/${FEATURES_ENVIRONMENT}/${feature}/
+  if [[ ! -d $feature_dir ]]; then
+    echo "[WARN] Feature '$feature' is not configured for environment '$FEATURES_ENVIRONMENT', skipping it"
+    continue
+  fi
+
+  echo "[INFO] Deleting feature '$feature' for environment '$FEATURES_ENVIRONMENT'"
+  set +e
+  if ! ${OC_TOOL} delete -k $feature_dir; then
```

Comment:
This might not work like we would think. Deletion can be tricky.

Deleting all the manifests for a feature doesn't necessarily mean the resource is cleaned up properly.

For example, deleting an operator, it's olm subscription, and it's custom resource all at once means we'd be potentially be stripping away the controller loop that would normally be handling tearing down the custom resource.

Also, it's possible to delete something like the "performance addon operator" immediately, but that doesn't mean the cluster has stabilized and actually removed that feature from the node. Each node in the MachineConfigPool will have to be updated with a new rendered MachineConfig before the deletion is 100% in place.

So, just like we have the "wait" functions after deploy, we'd likely need something like that during delete to guarantee the deletion has complete.

Reply:
> This might not work like we would think. Deletion can be tricky.
> 
> Deleting all the manifests for a feature doesn't necessarily mean the resource is cleaned up properly.
> 
> For example, deleting an operator, it's olm subscription, and it's custom resource all at once means we'd be potentially be stripping away the controller loop that would normally be handling tearing down the custom resource.

Agree, it depends on the purpouse of the operator. I think that it would work if the operator "just" produces new crs, as they are going to be owned by the "primary" resource. It might not be the case with the SRIOV operator which interacts with the underlying hw, for example. Once the operator is suddenly removed, the hw configuration will be the last applied one.


> 
> Also, it's possible to delete something like the "performance addon operator" immediately, but that doesn't mean the cluster has stabilized and actually removed that feature from the node. Each node in the MachineConfigPool will have to be updated with a new rendered MachineConfig before the deletion is 100% in place.
> 
> So, just like we have the "wait" functions after deploy, we'd likely need something like that during delete to guarantee the deletion has complete.

Again, this depends on the purpose. If we need this for manual clean up it's probably fine to observe it from outside. If not, we need the waiting logic in place.

So, it all boils down to: @cwilkers , what is the purpose of this new rule? What do you need it for?

Reply:
just fyi: this is what we do for the performance operator alone: https://github.com/openshift-kni/performance-addon-operators/blob/master/hack/clean-deploy.sh

Reply:
All good points... For my purpose, I was intending to simply clean up old versions of operators before retesting with newer, but it makes sense to have a `make clean` type option with greater scope, one which may try for removing node labels and waiting for updates to finish before proceeding with MCP then resource deletion.

In the case of something like SR-IOV where the hardware is not in a managed state without an operator present, there's not really a base configuration to wait on. If anything, it may be expected to delete anything using it first, then scale the virtual functions to zero.

On the other hand, if there is an expectation that simply calling on k8s to delete a feature will result in an unclean state, this is where Finalizers would come in to play. I agree that an operator should not wait on nodes to revert to a known state first because there are so many other factors at play there, but they should wait until their CRs are taken care of first, and CRs which have a direct relationship to hardware should leave that hardware in some predetermined state if removed. (Caveat, I'll admit I'm a bit out of my usual depth here, and may be proclaiming a number of 'should' statements which may not hold up to inspection...)